import { isTruthy } from './boolean';
import { typedHasProperty } from './object/typed-has-property';
import { isPromiseLike } from './promise/promise';
export function combineErrors(errors) {
    if (!errors || errors.length === 0) {
        return undefined;
    }
    const firstError = errors[0];
    if (errors.length === 1 && firstError) {
        return firstError;
    }
    return new Error(errors.map((error) => extractErrorMessage(error).trim()).join('\n'));
}
export function combineErrorMessages(errors) {
    if (!errors) {
        return '';
    }
    return errors.map(extractErrorMessage).filter(isTruthy).join('\n');
}
export function extractErrorMessage(maybeError) {
    if (!maybeError) {
        return '';
    }
    if (maybeError instanceof Error) {
        return maybeError.message;
    }
    else if (typedHasProperty(maybeError, 'message')) {
        return String(maybeError.message);
    }
    else {
        return String(maybeError);
    }
}
export function ensureError(maybeError) {
    if (maybeError instanceof Error) {
        return maybeError;
    }
    else {
        return new Error(extractErrorMessage(maybeError));
    }
}
export function ensureErrorAndPrependMessage(maybeError, prependMessage) {
    const error = ensureError(maybeError);
    error.message = `${prependMessage}: ${error.message}`;
    return error;
}
export function wrapInTry(inputs) {
    try {
        const returnValue = inputs.callback();
        if (returnValue instanceof Promise) {
            return returnValue.catch((error) => {
                if (inputs.catchCallback) {
                    return inputs.catchCallback(error);
                }
                else {
                    return inputs.fallbackValue;
                }
            });
        }
        else {
            return returnValue;
        }
    }
    catch (error) {
        if (inputs.catchCallback) {
            return inputs.catchCallback(error);
        }
        else {
            return inputs.fallbackValue;
        }
    }
}
export function executeAndReturnError(callback) {
    let caughtError;
    try {
        const result = callback();
        if (isPromiseLike(result)) {
            return new Promise(async (resolve) => {
                try {
                    const output = await result;
                    return resolve(output);
                }
                catch (error) {
                    caughtError = ensureError(error);
                }
                return resolve(caughtError);
            });
        }
        else {
            return result;
        }
    }
    catch (error) {
        caughtError = ensureError(error);
    }
    return caughtError;
}
